File: backend/app/schemas/scenario.py
from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field

from app.schemas.legal import LegalConceptRead
from app.schemas.world import LocationRead, SceneArtifactInput, SceneArtifactRead


class ScenarioGraphCreate(BaseModel):
    project_id: str
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    root_scene_id: Optional[str] = None


class ScenarioGraphUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    root_scene_id: Optional[str] = None


class ScenarioGraphRead(BaseModel):
    model_config = {'from_attributes': True}
    
    id: str
    project_id: str
    title: str
    description: Optional[str]
    root_scene_id: Optional[str]
    scenes: List["SceneNodeRead"] = Field(default_factory=list)
    edges: List["EdgeRead"] = Field(default_factory=list)


class SceneNodeCreate(BaseModel):
    title: str
    content: str
    synopsis: Optional[str] = None
    scene_type: Optional[str] = Field("story", pattern="^(story|decision)$")
    order_index: Optional[int] = None
    context: Optional[dict] = None
    location_id: Optional[str] = None
    location_material_set_id: Optional[str] = None
    location_overrides: Optional[dict] = None
    artifacts: Optional[List[SceneArtifactInput]] = None
    legal_concept_ids: Optional[List[str]] = None


class SceneNodeUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    synopsis: Optional[str] = None
    scene_type: Optional[str] = Field(None, pattern="^(story|decision)$")
    order_index: Optional[int] = None
    context: Optional[dict] = None
    location_id: Optional[str] = None
    location_material_set_id: Optional[str] = None
    location_overrides: Optional[dict] = None
    artifacts: Optional[List[SceneArtifactInput]] = None
    legal_concept_ids: Optional[List[str]] = None


class SceneNodeRead(BaseModel):
    model_config = {'from_attributes': True}
    
    id: str
    graph_id: str
    location_id: Optional[str]
    location_material_set_id: Optional[str]
    title: str
    content: str
    synopsis: Optional[str]
    scene_type: str
    order_index: Optional[int]
    context: Optional[dict]
    location_overrides: Optional[dict]
    location: Optional[LocationRead] = None
    artifacts: List[SceneArtifactRead] = Field(default_factory=list)
    legal_concepts: List[LegalConceptRead] = Field(default_factory=list)


class EdgeCreate(BaseModel):
    from_scene_id: str
    to_scene_id: str
    condition: Optional[str] = None
    choice_label: Optional[str] = None
    edge_metadata: Optional[dict] = None


class EdgeRead(BaseModel):
    model_config = {'from_attributes': True}
    
    id: str
    graph_id: str
    from_scene_id: str
    to_scene_id: str
    condition: Optional[str]
    choice_label: Optional[str]
    edge_metadata: Optional[dict]


class EdgeUpdate(BaseModel):
    choice_label: Optional[str] = None
    condition: Optional[str] = None
    edge_metadata: Optional[dict] = None


class GraphValidationIssue(BaseModel):
    code: str
    severity: str
    message: str
    scene_id: Optional[str] = None
    edge_id: Optional[str] = None
    metadata: Optional[dict] = None


class GraphValidationReport(BaseModel):
    graph_id: str
    issues: List[GraphValidationIssue] = Field(default_factory=list)
    summary: dict = Field(default_factory=dict)


class SceneUsageItem(BaseModel):
    scene_id: str
    title: str
    scene_type: str
    reason: str


class SceneUsageResponse(BaseModel):
    items: List[SceneUsageItem] = Field(default_factory=list)


ScenarioGraphRead.update_forward_refs()
SceneNodeRead.update_forward_refs()

File: frontend/src/api/scenario.ts
import { apiClient } from "./client";
import type { Edge, ScenarioGraph, SceneNode, GraphValidationReport, SceneUsageResponse } from "../shared/types";

export async function createGraph(projectId: string, payload: { title: string; description?: string | null }) {
  const response = await apiClient.post<ScenarioGraph>(`/v1/projects/${projectId}/graphs`, {
    project_id: projectId,
    ...payload,
  });
  return response.data;
}

export async function getGraph(graphId: string): Promise<ScenarioGraph> {
  const response = await apiClient.get<ScenarioGraph>(`/v1/graphs/${graphId}`);
  return response.data;
}

export async function updateGraph(
  graphId: string,
  payload: Partial<{ title: string; description?: string | null; root_scene_id?: string | null }>,
): Promise<ScenarioGraph> {
  const response = await apiClient.patch<ScenarioGraph>(`/v1/graphs/${graphId}`, payload);
  return response.data;
}

export async function deleteGraph(graphId: string): Promise<void> {
  await apiClient.delete(`/v1/graphs/${graphId}`);
}

export async function createScene(
  graphId: string,
  payload: {
    title: string;
    content: string;
    synopsis?: string | null;
    scene_type?: "story" | "decision";
    location_id?: string | null;
    location_overrides?: Record<string, unknown> | null;
    artifacts?: { artifact_id: string; state?: string | null; notes?: string | null; importance?: number }[];
  },
): Promise<SceneNode> {
  const response = await apiClient.post<SceneNode>(`/v1/graphs/${graphId}/scenes`, payload);
  return response.data;
}

export async function updateScene(
  sceneId: string,
  payload: Partial<{
    title: string;
    content: string;
    synopsis?: string | null;
    scene_type?: "story" | "decision";
    context?: Record<string, unknown> | null;
    legal_concept_ids?: string[];
    location_id?: string | null;
    location_overrides?: Record<string, unknown> | null;
    order_index?: number | null;
    artifacts?: { artifact_id: string; state?: string | null; notes?: string | null; importance?: number }[];
  }>,
): Promise<SceneNode> {
  const response = await apiClient.patch<SceneNode>(`/v1/scenes/${sceneId}`, payload);
  return response.data;
}

export async function createEdge(
  graphId: string,
  payload: {
    from_scene_id: string;
    to_scene_id: string;
    choice_label?: string | null;
    condition?: string | null;
    edge_metadata?: Record<string, unknown> | null;
  },
): Promise<Edge> {
  const response = await apiClient.post<Edge>(`/v1/graphs/${graphId}/edges`, payload);
  return response.data;
}

export async function updateEdge(
  edgeId: string,
  payload: {
    choice_label?: string | null;
    condition?: string | null;
    edge_metadata?: Record<string, unknown> | null;
  },
): Promise<Edge> {
  const response = await apiClient.patch<Edge>(`/v1/edges/${edgeId}`, payload);
  return response.data;
}

export async function validateGraph(graphId: string): Promise<GraphValidationReport> {
  const response = await apiClient.get<GraphValidationReport>(`/v1/graphs/${graphId}/validate`);
  return response.data;
}

export async function getSceneUsage(
  graphId: string,
  params: { location_id?: string; character_id?: string; artifact_id?: string },
): Promise<SceneUsageResponse> {
  const response = await apiClient.get<SceneUsageResponse>(`/v1/graphs/${graphId}/usage`, { params });
  return response.data;
}
